# LIBRAIRIES

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
import pickle

# DATASET

In [2]:
df = pd.read_csv("/Users/adamaboye/Downloads/PROJET TECH/projet_technique-G2/data/steam-games.csv")

In [3]:
df.head()

app_id                             title  release_date  \
0      730                  Counter-Strike 2  21 Aug, 2012   
1      570                            Dota 2   9 Jul, 2013   
2  2215430  Ghost of Tsushima DIRECTOR'S CUT  16 May, 2024   
3  1245620                        ELDEN RING  24 Feb, 2022   
4  1085660                         Destiny 2   1 Oct, 2019   

                            genres  \
0             Action, Free to Play   
1   Action, Strategy, Free to Play   
2                Action, Adventure   
3                      Action, RPG   
4  Action, Adventure, Free to Play   

                                          categories  \
0  Cross-Platform Multiplayer, Steam Trading Card...   
1  Steam Trading Cards, Steam Workshop, SteamVR C...   
2  Single-player, Online Co-op, Steam Achievement...   
3  Single-player, Online PvP, Online Co-op, Steam...   
4  Single-player, Online PvP, Online Co-op, Steam...   

                  developer           publisher original_price  \
0                     Valve               Valve            NaN   
1                     Valve               Valve            NaN   
2  Sucker Punch Productions  PlayStation PC LLC            NaN   
3         FromSoftware Inc.   FromSoftware Inc.            NaN   
4                    Bungie             Bungie             NaN   

  discount_percentage discounted_price  ...  win_support  mac_support  \
0                 NaN             Free  ...         True        False   
1                 NaN             Free  ...         True         True   
2                 NaN        ₹3,999.00  ...         True        False   
3                 NaN        ₹3,599.00  ...         True        False   
4                 NaN             Free  ...         True        False   

  linux_support awards  overall_review  overall_review_%  \
0          True      1   Very Positive              87.0   
1          True      0   Very Positive              81.0   
2         False      0   Very Positive              89.0   
3         False      6   Very Positive              93.0   
4         False      0   Very Positive              80.0   

   overall_review_count    recent_review recent_review_%  recent_review_count  
0             8062218.0  Mostly Positive            79.0              57466.0  
1             2243112.0  Mostly Positive            72.0              23395.0  
2               12294.0              NaN             NaN                  NaN  
3              605191.0    Very Positive            94.0               7837.0  
4              594713.0  Mostly Positive            73.0               4845.0  

[5 rows x 24 columns]

In [4]:
df.dtypes

app_id                    int64
title                    object
release_date             object
genres                   object
categories               object
developer                object
publisher                object
original_price           object
discount_percentage      object
discounted_price         object
dlc_available             int64
age_rating                int64
content_descriptor       object
about_description        object
win_support                bool
mac_support                bool
linux_support              bool
awards                    int64
overall_review           object
overall_review_%        float64
overall_review_count    float64
recent_review            object
recent_review_%         float64
recent_review_count     float64
dtype: object

In [5]:
genres = df["genres"].unique()
genres

array(['Action, Free to Play', 'Action, Strategy, Free to Play',
       'Action, Adventure', ...,
       'Action, Adventure, Casual, Racing, RPG, Simulation',
       'Action, Adventure, Sports, Early Access',
       'Massively Multiplayer, Racing, Early Access'], dtype=object)

In [6]:
list_genres = []

In [7]:
for genre in genres:
    if isinstance(genre, str):
        list_genres.extend(genre.split(', '))

In [8]:
list_genres = list(set(list_genres))

In [9]:
list_genres

['Racing',
 'Game Development',
 'Free to Play',
 'Education',
 'RPG',
 'Design & Illustration',
 'Sports',
 'Strategy',
 'Indie',
 'Early Access',
 'Web Publishing',
 'Video Production',
 'Utilities',
 'Casual',
 'Massively Multiplayer',
 'Animation & Modeling',
 'Movie',
 'Software Training',
 'Adventure',
 'Action',
 'Simulation',
 'Audio Production']

In [10]:
df['sport'] = df.apply(lambda row: int(any('sport' in str(x).lower() for x in row)), axis=1)
df['education'] = df.apply(lambda row: int(any('education' in str(x).lower() for x in row)), axis=1)
df['free_play'] = df.apply(lambda row: int(any('Free to Play' in str(x).lower() for x in row)), axis=1)
df['animation_modeling'] = df.apply(lambda row: int(any('Animation & Modeling' in str(x).lower() for x in row)), axis=1)
df['action'] = df.apply(lambda row: int(any('Action' in str(x).lower() for x in row)), axis=1)
df['simulation'] = df.apply(lambda row: int(any('Simulation' in str(x).lower() for x in row)), axis=1)
df['racing'] = df.apply(lambda row: int(any('Racing' in str(x).lower() for x in row)), axis=1)
df['video_production'] = df.apply(lambda row: int(any('Video Production' in str(x).lower() for x in row)), axis=1)

In [11]:
df['original_price'] = df['original_price'].str.replace('₹', '')
df['original_price'] = df['original_price'].str.replace(',', '')
df['original_price'] = df['original_price'].astype(float)

In [12]:
content_descriptor = df["content_descriptor"].unique()
content_descriptor

array([nan, 'Violence', 'Violence, In game purchases',
       'Violence, Sex, Bad Language', 'Violence, Bad Language, Sex',
       'Violence, In-Game Purchases',
       'Violence, Online Play,, Strong Language',
       'VIOLENCE, IN-GAME PURCHASES',
       'Violence, Bad Language, Gambling, In-Game Purchases (Includes Random Items)',
       'Violence, Online Gameplay', 'Extreme violence',
       'Violence, In-Game Purchases (Includes Random Items)',
       'Violence, Bad Language, In-Game Purchases',
       'In-Game Purchases(Includes Random Items)',
       'Violence, Drugs, Online Gameplay', 'Bad Language, Violence',
       'Violence, Users Interact, In-Game Purchases',
       'Violence, Bad Language', 'Violence, Bad language',
       'Fear, In-Game Purchases, In-Game Purchases (Includes Random Items)',
       'Violence, In-game Purchases (Includes Random Items)',
       'Fear, Mild Violence', 'In-Game Purchases',
       'Strong Language, Violence, Gambling', 'Sex, Violence',
       '

In [13]:
df['violence'] = df.apply(lambda row: int(any('violence' in str(x).lower() for x in row)), axis=1)
df['sex'] = df.apply(lambda row: int(any('sex' in str(x).lower() for x in row)), axis=1)
df['drugs'] = df.apply(lambda row: int(any('drugs' in str(x).lower() for x in row)), axis=1)
df['bad'] = df.apply(lambda row: int(any('sport' in str(x).lower() for x in row)), axis=1)
df['blood'] = df.apply(lambda row: int(any('blood' in str(x).lower() for x in row)), axis=1)

In [14]:
df

app_id                                              title  \
0          730                                   Counter-Strike 2   
1          570                                             Dota 2   
2      2215430                   Ghost of Tsushima DIRECTOR'S CUT   
3      1245620                                         ELDEN RING   
4      1085660                                          Destiny 2   
...        ...                                                ...   
42492   477910                               Wanderer of Teandria   
42493  1501390                                          KING PONG   
42494  2683250                              Falnarion Tactics III   
42495  1508840  Great Exploration VR: New Colony beyond Viking...   
42496  1191080                                           SANCTION   

       release_date                                             genres  \
0      21 Aug, 2012                               Action, Free to Play   
1       9 Jul, 2013                     Action, Strategy, Free to Play   
2      16 May, 2024                                  Action, Adventure   
3      24 Feb, 2022                                        Action, RPG   
4       1 Oct, 2019                    Action, Adventure, Free to Play   
...             ...                                                ...   
42492  26 Sep, 2017                           Action, Adventure, Indie   
42493  18 Feb, 2021                       Action, Sports, Early Access   
42494  25 Dec, 2023                                           Strategy   
42495  12 Jan, 2021                                     Action, Casual   
42496  10 Jun, 2022  Action, Adventure, Casual, Indie, Simulation, ...   

                                              categories  \
0      Cross-Platform Multiplayer, Steam Trading Card...   
1      Steam Trading Cards, Steam Workshop, SteamVR C...   
2      Single-player, Online Co-op, Steam Achievement...   
3      Single-player, Online PvP, Online Co-op, Steam...   
4      Single-player, Online PvP, Online Co-op, Steam...   
...                                                  ...   
42492  Single-player, Steam Achievements, Steam Tradi...   
42493  Single-player, Online PvP, Steam Achievements,...   
42494  Single-player, Steam Achievements, Steam Cloud...   
42495  Single-player, Tracked Controller Support, VR ...   
42496  Single-player, Steam Achievements, Family Shar...   

                      developer             publisher  original_price  \
0                         Valve                 Valve             NaN   
1                         Valve                 Valve             NaN   
2      Sucker Punch Productions    PlayStation PC LLC             NaN   
3             FromSoftware Inc.     FromSoftware Inc.             NaN   
4                        Bungie               Bungie              NaN   
...                         ...                   ...             ...   
42492           Silentplaygames       Silentplaygames           250.0   
42493                    Iconik                Iconik           529.0   
42494              Team Syukino          Team Syukino             NaN   
42495         William at Oxford     William at Oxford           199.0   
42496      LethalLizard Studios  LethalLizard Studios             NaN   

      discount_percentage discounted_price  ...  animation_modeling  action  \
0                     NaN             Free  ...                   0       0   
1                     NaN             Free  ...                   0       0   
2                     NaN        ₹3,999.00  ...                   0       0   
3                     NaN        ₹3,599.00  ...                   0       0   
4                     NaN             Free  ...                   0       0   
...                   ...              ...  ...                 ...     ...   
42492                -50%          ₹125.00  ...                   0       0   
42493                -50%          ₹264.00  ...             

In [15]:
X = df[["sport", "education", "free_play", "animation_modeling", "action", "simulation", "racing", "video_production", "violence", "sex", "drugs", "bad", "blood"]]
Y = df["age_rating"]

In [16]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

# MODELES

1) Linear Regression

In [17]:
model_lr = LinearRegression()

In [18]:
model_lr.fit(X_train, Y_train)

LinearRegression()

In [19]:
print('Coefficients:', model_lr.coef_)

Coefficients: [ 2.48791034e+08  5.49969077e-03 -1.23890020e+01  9.09273042e-01
 -1.61565585e+01 -1.14067049e+01 -1.70990010e-03  0.00000000e+00
  9.43712670e-01  2.53953998e-01  1.17645789e-01 -2.48791034e+08
  1.18958398e-03]


In [20]:
print('Intercept:', model_lr.intercept_)

Intercept: 0.008990916046654314


In [21]:
Y_pred = model_lr.predict(X_test)

In [22]:
mse_lr = mean_squared_error(Y_test, Y_pred)
mse_lr

0.012375565576335033

In [23]:
r2 = r2_score(Y_test, Y_pred)
r2

0.7760463777473192

In [24]:
with open('video_game_lr_model.pkl', 'wb') as file:
    pickle.dump(model_lr, file)

2) Sequential

In [25]:
model_seq = Sequential()

In [26]:
model_seq.add(Dense(10, input_dim=X_train.shape[1], activation='relu'))
model_seq.add(Dense(1, activation='linear'))

/opt/homebrew/anaconda3/lib/python3.11/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [27]:
model_seq.compile(loss='mean_squared_error', optimizer='adam', metrics=['mse'])

In [28]:
history = model_seq.fit(X_train, Y_train, epochs=15, batch_size=32, validation_data=(X_test, Y_test))

Epoch 1/15
1063/1063 ━━━━━━━━━━━━━━━━━━━━ 1s 583us/step - loss: 0.0275 - mse: 0.0275 - val_loss: 0.0124 - val_mse: 0.0124
Epoch 2/15
1063/1063 ━━━━━━━━━━━━━━━━━━━━ 0s 422us/step - loss: 0.0119 - mse: 0.0119 - val_loss: 0.0125 - val_mse: 0.0125
Epoch 3/15
1063/1063 ━━━━━━━━━━━━━━━━━━━━ 0s 451us/step - loss: 0.0121 - mse: 0.0121 - val_loss: 0.0123 - val_mse: 0.0123
Epoch 4/15
1063/1063 ━━━━━━━━━━━━━━━━━━━━ 0s 424us/step - loss: 0.0108 - mse: 0.0108 - val_loss: 0.0123 - val_mse: 0.0123
Epoch 5/15
1063/1063 ━━━━━━━━━━━━━━━━━━━━ 0s 437us/step - loss: 0.0111 - mse: 0.0111 - val_loss: 0.0122 - val_mse: 0.0122
Epoch 6/15
1063/1063 ━━━━━━━━━━━━━━━━━━━━ 0s 418us/step - loss: 0.0118 - mse: 0.0118 - val_loss: 0.0123 - val_mse: 0.0123
Epoch 7/15
1063/1063 ━━━━━━━━━━━━━━━━━━━━ 0s 438us/step - loss: 0.0106 - mse: 0.0106 - val_loss: 0.0123 - val_mse: 0.0123
Epoch 8/15
1063/1063 ━━━━━━━━━━━━━━━━━━━━ 0s 451us/step - loss: 0.0116 - mse: 0.0116 - val_loss: 0.0122 - val_mse: 0.0122
Epoch 9/15
1063/1063 ━━━

In [29]:
loss_seq, mse_seq = model_seq.evaluate(X_test, Y_test)

266/266 ━━━━━━━━━━━━━━━━━━━━ 0s 277us/step - loss: 0.0113 - mse: 0.0113      


In [30]:
loss_seq

0.012243784964084625

In [31]:
mse_seq

0.012243784964084625

In [32]:
with open('video_game_seq_model.pkl', 'wb') as file:
    pickle.dump(model_seq, file)

# COMPARAISON MODELES

Performance de la régression linéaire
- Mean Squared Error (MSE) : 0.0123
- R-squared (R²) : 0.78

Performance du modèle séquentiel (réseau de neurones)
- Mean Squared Error (MSE) : 0.0122

# CHOIX DU MODELE 

En termes de MSE, les deux modèles semblent avoir des performances assez similaires sur votre ensemble de données de test. 

Mais on choisie le modèle Linear Regression car le Sequential est une boite noire et plus complexe à expliquer